File: Pastura/Pastura/Engine/ConditionEvaluator.swift
import Foundation

/// Evaluates a boolean condition expression used by the `conditional` phase
/// type. Supports `&&` / `||` combinators and parenthesized grouping on top
/// of a single-comparison primitive.
///
/// Grammar:
///
///     expression ::= or
///     or         ::= and ('||' and)*
///     and        ::= factor ('&&' factor)*
///     factor     ::= '(' or ')'  |  comparison
///     comparison ::= operand OP operand
///     OP         ::= "==" | "!=" | "<=" | ">=" | "<" | ">"
///     operand    ::= Identifier ("." Identifier)?
///                    | NumberLiteral
///                    | StringLiteral
///     StringLiteral ::= '"' .* '"'
///
/// Precedence (loosest to tightest): `||` < `&&` < comparison. Both `&&`
/// and `||` are left-associative — `a || b || c` evaluates as
/// `(a || b) || c`. Parens may group any boolean sub-expression and
/// override precedence; an operand itself cannot be parenthesized
/// (`(current_round) == 1` is rejected).
///
/// Tokenization is quote-aware: operator-like characters inside a quoted
/// string literal (e.g. `"a && b"`, `"x>y"`, `"(foo)"`) are treated as
/// literal content rather than tokens.
///
/// Derived read-only variables available on either side of a comparison:
///
/// | Identifier         | Source                                      |
/// |--------------------|---------------------------------------------|
/// | `current_round`    | `state.currentRound`                        |
/// | `total_rounds`     | `scenario.rounds`                           |
/// | `max_score`        | `state.scores.values.max()`                 |
/// | `min_score`        | `state.scores.values.min()`                 |
/// | `eliminated_count` | count of `state.eliminated.values == true`  |
/// | `active_count`     | count of `state.eliminated.values == false` |
/// | `vote_winner`      | most-voted name in `state.voteResults` (ties broken by count desc, name desc) |
/// | `scores.<Name>`    | `state.scores["<Name>"]`                    |
///
/// Any other identifier is resolved from `state.variables`.
///
/// **Parse-time errors** (missing operator, empty operand, mismatched
/// parens, dangling `&&` / `||`, empty `()`) throw
/// `SimulationError.scenarioValidationFailed`. Use ``parse(_:)`` to fail
/// fast at scenario-load time rather than waiting for handler dispatch
/// — `ScenarioValidator.validateConditionalPhase` does this.
///
/// **Runtime-absent values** (e.g. `vote_winner` before any vote this
/// round, `scores.Nobody`) do not throw; the comparison they appear in
/// returns `value: false` and a warning string is appended to
/// ``EvaluationResult/warnings`` so the caller can surface it via the
/// normal `.summary` warning channel.
///
/// **Short-circuit policy** (Swift-style): the dropped side of `false &&
/// X` / `true || X` is **not** resolved, so absent-variable warnings on
/// the skipped branch never appear. Trade-off: a typo in a never-evaluated
/// sub-expression (e.g. `current_round > 999 && scores.Aliec > 5` with
/// `Alice` typo'd) stays hidden because the always-false LHS short-
/// circuits before the parser ever asks the resolver about `Aliec`.
/// Authors debugging an always-same-branch condition should temporarily
/// flip operands or remove the combinator to surface warnings on both
/// sides.
///
/// This type is the sole owner of the expression grammar — callers pass an
/// expression string only. Tokenizer + recursive-descent parser internals
/// live in `ConditionEvaluator+Parser.swift`.
nonisolated public struct ConditionEvaluator: Sendable {

  /// Result of evaluating a condition expression.
  public struct EvaluationResult: Sendable, Equatable {
    /// The evaluated boolean. `false` when a comparison's operand is
    /// runtime-absent, or when a sub-expression's truth value (and short-
    /// circuit policy) drives it false.
    public let value: Bool

    /// Non-fatal diagnostics (e.g. runtime-absent variables). Callers
    /// should forward these to the `.summary` event so users can debug
    /// their DSL. Warnings from short-circuited sub-expressions are
    /// suppressed by design — see the type-level doc comment.
    public let warnings: [String]
  }

  /// Lexical token classes recognized by the DSL. Defined as a nested
  /// type so the parser file (a sibling extension) can reference it
  /// without widening visibility module-internal-or-tighter.
  enum Token: Equatable {
    case openParen
    case closeParen
    case logicalAnd
    case logicalOr
    case comparisonOp(String)
    /// String preserving its source form: numbers as their digits, bare
    /// identifiers as-is (including dotted access like `scores.Alice`),
    /// and quoted strings WITH surrounding `"` so the resolver can
    /// distinguish `Alice` (identifier) from `"Alice"` (string literal).
    case operand(String)
  }

  /// AST node for the parsed expression. Walker lives in this file;
  /// builder lives in `ConditionEvaluator+Parser.swift`.
  indirect enum Node: Equatable {
    case logicalOr(Node, Node)
    case logicalAnd(Node, Node)
    case comparison(lhs: String, symbol: String, rhs: String)
  }

  public init() {}

  /// Evaluates `expression` against `state` and `scenario`.
  ///
  /// - Throws: `SimulationError.scenarioValidationFailed` for parse-time
  ///   errors (missing operator, empty operand, mismatched parens,
  ///   dangling combinator, empty parens).
  public func evaluate(
    _ expression: String,
    state: SimulationState,
    scenario: Scenario
  ) throws -> EvaluationResult {
    let ast = try parseToAST(expression)
    return walk(ast, state: state, scenario: scenario)
  }

  /// Parses `expression` and discards the AST. Used by
  /// `ScenarioValidator.validateConditionalPhase` to surface malformed
  /// `if:` strings at scenario-load time, before any `state` exists.
  ///
  /// - Throws: `SimulationError.scenarioValidationFailed` on syntactic
  ///   errors. Runtime-absent identifiers do not throw — those only
  ///   surface during ``evaluate(_:state:scenario:)`` as warnings.
  public func parse(_ expression: String) throws {
    _ = try parseToAST(expression)
  }

  // MARK: - AST walker (with short-circuit)

  private func walk(
    _ node: Node, state: SimulationState, scenario: Scenario
  ) -> EvaluationResult {
    switch node {
    case .comparison(let lhs, let symbol, let rhs):
      return walkComparison(
        lhs: lhs, symbol: symbol, rhs: rhs, state: state, scenario: scenario)
    case .logicalAnd(let lhs, let rhs):
      let lhsResult = walk(lhs, state: state, scenario: scenario)
      // Short-circuit: dropped side is NOT resolved, so its warnings
      // never enter the result. Documented Swift-style policy.
      if !lhsResult.value { return lhsResult }
      let rhsResult = walk(rhs, state: state, scenario: scenario)
      return EvaluationResult(
        value: rhsResult.value,
        warnings: lhsResult.warnings + rhsResult.warnings)
    case .logicalOr(let lhs, let rhs):
      let lhsResult = walk(lhs, state: state, scenario: scenario)
      if lhsResult.value { return lhsResult }
      let rhsResult = walk(rhs, state: state, scenario: scenario)
      return EvaluationResult(
        value: rhsResult.value,
        warnings: lhsResult.warnings + rhsResult.warnings)
    }
  }

  private func walkComparison(
    lhs: String, symbol: String, rhs: String,
    state: SimulationState, scenario: Scenario
  ) -> EvaluationResult {
    var warnings: [String] = []
    let lhsValue = resolve(
      token: lhs, state: state, scenario: scenario, warnings: &warnings)
    let rhsValue = resolve(
      token: rhs, state: state, scenario: scenario, warnings: &warnings)
    guard let left = lhsValue, let right = rhsValue else {
      return EvaluationResult(value: false, warnings: warnings)
    }
    return EvaluationResult(
      value: compare(left, symbol, right), warnings: warnings)
  }

  // MARK: - Operand resolution

  /// Resolves a token to a string value, or returns `nil` if the
  /// identifier refers to data not present at runtime (with a warning
  /// appended to `warnings`).
  private func resolve(
    token: String,
    state: SimulationState,
    scenario: Scenario,
    warnings: inout [String]
  ) -> String? {
    if token.hasPrefix("\"") && token.hasSuffix("\"") && token.count >= 2 {
      return String(token.dropFirst().dropLast())
    }
    if Double(token) != nil {
      return token
    }
    if let dotIndex = token.firstIndex(of: ".") {
      return resolveDotted(token, dotIndex: dotIndex, state: state, warnings: &warnings)
    }

    switch resolveDerived(token, state: state, scenario: scenario, warnings: &warnings) {
    case .value(let resolved):
      return resolved
    case .absent:
      return nil
    case .notDerived:
      break
    }

    if let fromVariables = state.variables[token] {
      return fromVariables
    }
    warnings.append("Unknown identifier '\(token)'")
    return nil
  }

  private func resolveDotted(
    _ token: String,
    dotIndex: String.Index,
    state: SimulationState,
    warnings: inout [String]
  ) -> String? {
    let head = String(token[..<dotIndex])
    let tail = String(token[token.index(after: dotIndex)...])
    if head == "scores" {
      if let score = state.scores[tail] {
        return String(score)
      }
      warnings.append("scores.\(tail) is not set (agent absent from scores)")
      return nil
    }
    warnings.append("Unknown dotted identifier '\(token)'")
    return nil
  }

  /// Tri-state result for derived-variable resolution.
  private enum DerivedResolution {
    case value(String)
    /// Recognized as derived but has no runtime value. A warning has
    /// already been appended by the producer.
    case absent
    /// Not a derived variable at all; caller should fall through to
    /// `state.variables`.
    case notDerived
  }

  private func resolveDerived(
    _ identifier: String,
    state: SimulationState,
    scenario: Scenario,
    warnings: inout [String]
  ) -> DerivedResolution {
    if let value = resolveAlwaysPresentDerived(identifier, state: state, scenario: scenario) {
      return .value(value)
    }
    return resolveMayBeAbsentDerived(identifier, state: state, warnings: &warnings)
  }

  private func resolveAlwaysPresentDerived(
    _ identifier: String, state: SimulationState, scenario: Scenario
  ) -> String? {
    switch identifier {
    case "current_round": return String(state.currentRound)
    case "total_rounds": return String(scenario.rounds)
    case "eliminated_count": return String(state.eliminated.values.filter { $0 }.count)
    case "active_count": return String(state.eliminated.values.filter { !$0 }.count)
    default: return nil
    }
  }

  private func resolveMayBeAbsentDerived(
    _ identifier: String, state: SimulationState, warnings: inout [String]
  ) -> DerivedResolution {
    switch identifier {
    case "max_score":
      return resolveScoreExtremum(
        state.scores.values.max(), label: "max_score", warnings: &warnings)
    case "min_score":
      return resolveScoreExtremum(
        state.scores.values.min(), label: "min_score", warnings: &warnings)
    case "vote_winner":
      // Shared canonical tie-break (count desc, name desc). EliminateHandler
      // and WordwolfJudgeLogic resolve the same winner (#1056).
      if let winner = VoteTally.winner(state.voteResults) {
        return .value(winner.key)
      }
      warnings.append("vote_winner has no value (no vote phase has run this round)")
      return .absent
    default:
      return .notDerived
    }
  }

  private func resolveScoreExtremum(
    _ value: Int?, label: String, warnings: inout [String]
  ) -> DerivedResolution {
    if let value = value {
      return .value(String(value))
    }
    warnings.append("\(label) has no value (scores is empty)")
    return .absent
  }

  // MARK: - Comparison

  private func compare(_ lhs: String, _ symbol: String, _ rhs: String) -> Bool {
    if let lhsNum = Double(lhs), let rhsNum = Double(rhs) {
      return applyOperator(symbol, lhsNum, rhsNum)
    }
    return applyOperator(symbol, lhs, rhs)
  }

  private func applyOperator<T: Comparable>(_ symbol: String, _ lhs: T, _ rhs: T) -> Bool {
    switch symbol {
    case "==": return lhs == rhs
    case "!=": return lhs != rhs
    case "<": return lhs < rhs
    case "<=": return lhs <= rhs
    case ">": return lhs > rhs
    case ">=": return lhs >= rhs
    default: return false
    }
  }
}

File: shared/engine/src/commonMain/kotlin/com/pastura/engine/ConditionEvaluator.kt
package com.pastura.engine

import com.pastura.models.Scenario
import com.pastura.models.SimulationError
import com.pastura.models.SimulationState

/**
 * Evaluates a boolean condition expression used by the `conditional` phase
 * type. Supports `&&` / `||` combinators and parenthesized grouping on top of a
 * single-comparison primitive.
 *
 * Kotlin port of `Pastura/Pastura/Engine/ConditionEvaluator.swift` +
 * `ConditionEvaluator+Parser.swift` (#501 Stage 2-pre / ADR-023 §6). The Swift
 * original splits tokenizer/parser into a sibling `nonisolated extension` to
 * dodge a default-MainActor isolation trap; Kotlin has no such trap, so the
 * whole grammar (public API, walker, resolver, tokenizer, recursive-descent
 * parser) lives in this one file. The Swift `ConditionEvaluator` is NOT deleted —
 * iOS keeps using it; this port runs in parallel until the Stage-5 consumption
 * switch. `ConditionEvaluatorTests` (commonTest) is the cross-language executable
 * spec.
 *
 * Grammar:
 * ```
 * expression ::= or
 * or         ::= and ('||' and)*
 * and        ::= factor ('&&' factor)*
 * factor     ::= '(' or ')'  |  comparison
 * comparison ::= operand OP operand
 * OP         ::= "==" | "!=" | "<=" | ">=" | "<" | ">"
 * operand    ::= Identifier ("." Identifier)? | NumberLiteral | StringLiteral
 * StringLiteral ::= '"' .* '"'
 * ```
 *
 * Precedence (loosest to tightest): `||` < `&&` < comparison. Both are
 * left-associative. Parens group any boolean sub-expression; an operand itself
 * cannot be parenthesized (`(current_round) == 1` is rejected).
 *
 * Tokenization is quote-aware: operator-like characters inside a quoted string
 * literal (`"a && b"`, `"x>y"`, `"(foo)"`) are literal content, not tokens.
 *
 * Derived read-only variables (either side of a comparison): `current_round`,
 * `total_rounds`, `max_score`, `min_score`, `eliminated_count`, `active_count`,
 * `vote_winner` (most-voted name, ties broken like `EliminateHandler`),
 * `scores.<Name>`. Any other identifier resolves from `state.variables`.
 *
 * **Parse-time errors** (missing operator, empty operand, mismatched/empty
 * parens, dangling `&&`/`||`, unterminated string) throw a [SimulationException]
 * carrying [SimulationError.ScenarioValidationFailed]. (Swift throws the sealed
 * `SimulationError` directly; Kotlin's is not a `Throwable`, so the Engine wraps
 * it — see [SimulationException].) Use [parse] to fail fast at scenario-load time.
 *
 * **Runtime-absent values** (e.g. `vote_winner` pre-vote, `scores.Nobody`) do NOT
 * throw; the comparison returns `value = false` and appends a warning to
 * [EvaluationResult.warnings].
 *
 * **Short-circuit policy** (Swift-style): the dropped side of `false && X` /
 * `true || X` is NOT resolved, so absent-variable warnings on the skipped branch
 * never appear.
 *
 * **Cross-language numeric parity.** Numeric detection uses a fixed decimal-literal
 * regex ([numericLiteralRegex]) rather than raw `String.toDoubleOrNull()`. Neither
 * raw parser is a stable cross-language predicate: Kotlin `toDoubleOrNull` accepts
 * `"Infinity"` / `"NaN"` / type-suffixed literals (`"1f"`) that are not plain
 * decimals, while Swift `Double(String)` accepts hex-floats (`"0x1p4"`) Kotlin
 * rejects. The regex accepts a token iff it is a plain optionally-signed decimal
 * with optional fraction / exponent — identical in Swift and Kotlin on that
 * grammar. Everything else (`Infinity` / `NaN` / suffixed `1f` — normalizing
 * Kotlin's laxer acceptance; hex-floats — the one Swift accepts) takes the string
 * path. All out-of-domain: real operands are integers or names. Pinned by
 * `ConditionEvaluatorParityTests`.
 */
public class ConditionEvaluator {

    /** Result of evaluating a condition expression. */
    public data class EvaluationResult(
        /**
         * The evaluated boolean. `false` when a comparison's operand is
         * runtime-absent, or when the sub-expression's truth value (and
         * short-circuit policy) drives it false.
         */
        public val value: Boolean,
        /**
         * Non-fatal diagnostics (e.g. runtime-absent variables). Warnings from
         * short-circuited sub-expressions are suppressed by design.
         */
        public val warnings: List<String>,
    )

    /** Lexical token classes recognized by the DSL. */
    internal sealed interface Token {
        data object OpenParen : Token
        data object CloseParen : Token
        data object LogicalAnd : Token
        data object LogicalOr : Token
        data class ComparisonOp(val symbol: String) : Token

        /**
         * String preserving its source form: numbers as their digits, bare
         * identifiers as-is (including dotted access like `scores.Alice`), and
         * quoted strings WITH surrounding `"` so the resolver can distinguish
         * `Alice` (identifier) from `"Alice"` (string literal).
         */
        data class Operand(val text: String) : Token
    }

    /** AST node for the parsed expression (Swift's `indirect enum`). */
    internal sealed interface Node {
        data class LogicalOr(val lhs: Node, val rhs: Node) : Node
        data class LogicalAnd(val lhs: Node, val rhs: Node) : Node
        data class Comparison(val lhs: String, val symbol: String, val rhs: String) : Node
    }

    /**
     * Evaluates [expression] against [state] and [scenario].
     *
     * @throws SimulationException carrying [SimulationError.ScenarioValidationFailed]
     *   for parse-time errors.
     */
    public fun evaluate(
        expression: String,
        state: SimulationState,
        scenario: Scenario,
    ): EvaluationResult {
        val ast = parseToAST(expression)
        return walk(ast, state, scenario)
    }

    /**
     * Parses [expression] and discards the AST. Used to surface malformed `if:`
     * strings at scenario-load time, before any `state` exists.
     *
     * @throws SimulationException on syntactic errors. Runtime-absent identifiers
     *   do NOT throw — those only surface during [evaluate] as warnings.
     */
    public fun parse(expression: String) {
        parseToAST(expression)
    }

    // MARK: - AST walker (with short-circuit)

    private fun walk(node: Node, state: SimulationState, scenario: Scenario): EvaluationResult =
        when (node) {
            is Node.Comparison ->
                walkComparison(node.lhs, node.symbol, node.rhs, state, scenario)

            is Node.LogicalAnd -> {
                val lhsResult = walk(node.lhs, state, scenario)
                // Short-circuit: dropped side is NOT resolved, so its warnings
                // never enter the result. Documented Swift-style policy.
                if (!lhsResult.value) {
                    lhsResult
                } else {
                    val rhsResult = walk(node.rhs, state, scenario)
                    EvaluationResult(rhsResult.value, lhsResult.warnings + rhsResult.warnings)
                }
            }

            is Node.LogicalOr -> {
                val lhsResult = walk(node.lhs, state, scenario)
                if (lhsResult.value) {
                    lhsResult
                } else {
                    val rhsResult = walk(node.rhs, state, scenario)
                    EvaluationResult(rhsResult.value, lhsResult.warnings + rhsResult.warnings)
                }
            }
        }

    private fun walkComparison(
        lhs: String,
        symbol: String,
        rhs: String,
        state: SimulationState,
        scenario: Scenario,
    ): EvaluationResult {
        val warnings = mutableListOf<String>()
        val lhsValue = resolve(lhs, state, scenario, warnings)
        val rhsValue = resolve(rhs, state, scenario, warnings)
        if (lhsValue == null || rhsValue == null) {
            return EvaluationResult(false, warnings)
        }
        return EvaluationResult(compare(lhsValue, symbol, rhsValue), warnings)
    }

    // MARK: - Operand resolution

    /**
     * Resolves a token to a string value, or `null` if the identifier refers to
     * data not present at runtime (with a warning appended to [warnings]).
     */
    private fun resolve(
        token: String,
        state: SimulationState,
        scenario: Scenario,
        warnings: MutableList<String>,
    ): String? {
        if (token.length >= 2 && token.startsWith("\"") && token.endsWith("\"")) {
            return token.substring(1, token.length - 1)
        }
        if (isNumericLiteral(token)) {
            return token
        }
        val dotIndex = token.indexOf('.')
        if (dotIndex >= 0) {
            return resolveDotted(token, dotIndex, state, warnings)
        }

        when (val derived = resolveDerived(token, state, scenario, warnings)) {
            is DerivedResolution.Value -> return derived.value
            DerivedResolution.Absent -> return null
            DerivedResolution.NotDerived -> Unit
        }

        state.variables[token]?.let { return it }
        warnings.add("Unknown identifier '$token'")
        return null
    }

    private fun resolveDotted(
        token: String,
        dotIndex: Int,
        state: SimulationState,
        warnings: MutableList<String>,
    ): String? {
        val head = token.substring(0, dotIndex)
        val tail = token.substring(dotIndex + 1)
        if (head == "scores") {
            state.scores[tail]?.let { return it.toString() }
            warnings.add("scores.$tail is not set (agent absent from scores)")
            return null
        }
        warnings.add("Unknown dotted identifier '$token'")
        return null
    }

    /** Tri-state result for derived-variable resolution. */
    private sealed interface DerivedResolution {
        data class Value(val value: String) : DerivedResolution

        /** Recognized as derived but has no runtime value (warning already appended). */
        data object Absent : DerivedResolution

        /** Not a derived variable at all; caller falls through to `state.variables`. */
        data object NotDerived : DerivedResolution
    }

    private fun resolveDerived(
        identifier: String,
        state: SimulationState,
        scenario: Scenario,
        warnings: MutableList<String>,
    ): DerivedResolution {
        resolveAlwaysPresentDerived(identifier, state, scenario)?.let {
            return DerivedResolution.Value(it)
        }
        return resolveMayBeAbsentDerived(identifier, state, warnings)
    }

    private fun resolveAlwaysPresentDerived(
        identifier: String,
        state: SimulationState,
        scenario: Scenario,
    ): String? = when (identifier) {
        "current_round" -> state.currentRound.toString()
        "total_rounds" -> scenario.rounds.toString()
        "eliminated_count" -> state.eliminated.values.count { it }.toString()
        "active_count" -> state.eliminated.values.count { !it }.toString()
        else -> null
    }

    private fun resolveMayBeAbsentDerived(
        identifier: String,
        state: SimulationState,
        warnings: MutableList<String>,
    ): DerivedResolution = when (identifier) {
        "max_score" ->
            resolveScoreExtremum(state.scores.values.maxOrNull(), "max_score", warnings)

        "min_score" ->
            resolveScoreExtremum(state.scores.values.minOrNull(), "min_score", warnings)

        "vote_winner" -> {
            // Deterministic tie-break mirrors EliminateHandler: sort by
            // (count desc, name desc) and take the first. Kotlin string order
            // (UTF-16) matches Swift scalar order across the BMP (agent names);
            // see the port's numeric/ordering-parity doc note and
            // ConditionEvaluatorParityTests. Future parity with a ported
            // EliminateHandler is a Stage-3 concern.
            val top = state.voteResults.entries
                .sortedWith(
                    compareByDescending<Map.Entry<String, Int>> { it.value }
                        .thenByDescending { it.key },
                )
                .firstOrNull()
            if (top != null) {
                DerivedResolution.Value(top.key)
            } else {
                warnings.add("vote_winner has no value (no vote phase has run this round)")
                DerivedResolution.Absent
            }
        }

        else -> DerivedResolution.NotDerived
    }

    private fun resolveScoreExtremum(
        value: Int?,
        label: String,
        warnings: MutableList<String>,
    ): DerivedResolution {
        if (value != null) {
            return DerivedResolution.Value(value.toString())
        }
        warnings.add("$label has no value (scores is empty)")
        return DerivedResolution.Absent
    }

    // MARK: - Comparison

    private fun compare(lhs: String, symbol: String, rhs: String): Boolean {
        val lhsNum = numericValue(lhs)
        val rhsNum = numericValue(rhs)
        if (lhsNum != null && rhsNum != null) {
            return applyOperator(symbol, lhsNum, rhsNum)
        }
        return applyOperator(symbol, lhs, rhs)
    }

    private fun applyOperator(symbol: String, lhs: Double, rhs: Double): Boolean = when (symbol) {
        "==" -> lhs == rhs
        "!=" -> lhs != rhs
        "<" -> lhs < rhs
        "<=" -> lhs <= rhs
        ">" -> lhs > rhs
        ">=" -> lhs >= rhs
        else -> false
    }

    private fun applyOperator(symbol: String, lhs: String, rhs: String): Boolean = when (symbol) {
        "==" -> lhs == rhs
        "!=" -> lhs != rhs
        "<" -> lhs < rhs
        "<=" -> lhs <= rhs
        ">" -> lhs > rhs
        ">=" -> lhs >= rhs
        else -> false
    }

    // MARK: - Tokenizer

    private fun tokenize(expression: String): List<Token> {
        val chars = expression
        val tokens = mutableListOf<Token>()
        var index = 0

        while (index < chars.length) {
            val char = chars[index]
            if (char.isWhitespace()) {
                index += 1
                continue
            }
            // Two-char operators scanned before single-char prefixes so `<=` is
            // not split into `<` + `=`, and `&&` / `||` are not read as bare chars.
            val multi = readMultiCharToken(chars, index)
            if (multi != null) {
                tokens.add(multi.first)
                index += multi.second
                continue
            }
            val single = readSingleCharToken(chars, index)
            if (single != null) {
                tokens.add(single.first)
                index += single.second
                continue
            }
            if (char == '"') {
                val quoted = readQuotedOperand(chars, index, expression)
                tokens.add(quoted.first)
                index = quoted.second
                continue
            }
            val bare = readBareOperand(chars, index, expression)
            tokens.add(bare.first)
            index = bare.second
        }

        return tokens
    }

    /** Two-character operators (`&&`, `||`, `==`, `!=`, `<=`, `>=`); `null` otherwise. */
    private fun readMultiCharToken(chars: String, index: Int): Pair<Token, Int>? {
        if (index + 1 >= chars.length) return null
        val pair = chars.substring(index, index + 2)
        return when (pair) {
            "&&" -> Token.LogicalAnd to 2
            "||" -> Token.LogicalOr to 2
            "==", "!=", "<=", ">=" -> Token.ComparisonOp(pair) to 2
            else -> null
        }
    }

    /** Single-character tokens (`(`, `)`, `<`, `>`); `null` otherwise. */
    private fun readSingleCharToken(chars: String, index: Int): Pair<Token, Int>? =
        when (chars[index]) {
            '(' -> Token.OpenParen to 1
            ')' -> Token.CloseParen to 1
            '<', '>' -> Token.ComparisonOp(chars[index].toString()) to 1
            else -> null
        }

    /** Quoted string operand. Keeps the surrounding `"` so the resolver can tell literals apart. */
    private fun readQuotedOperand(chars: String, index: Int, source: String): Pair<Token, Int> {
        var end = index + 1
        while (end < chars.length && chars[end] != '"') {
            end += 1
        }
        if (end >= chars.length) {
            throw SimulationException(
                SimulationError.ScenarioValidationFailed(
                    "Condition expression '$source' has unterminated string literal",
                ),
            )
        }
        return Token.Operand(chars.substring(index, end + 1)) to (end + 1)
    }

    /**
     * Bare operand: identifier, number, or dotted access. Accumulates until a
     * delimiter (whitespace, paren, quote, operator-prefix) is hit.
     */
    private fun readBareOperand(chars: String, index: Int, source: String): Pair<Token, Int> {
        var end = index
        while (end < chars.length && !isOperandDelimiter(chars, end)) {
            end += 1
        }
        if (end == index) {
            throw SimulationException(
                SimulationError.ScenarioValidationFailed(
                    "Condition expression '$source' contains unexpected character '${chars[index]}'",
                ),
            )
        }
        return Token.Operand(chars.substring(index, end)) to end
    }

    /** True when `chars[end]` starts a delimiter that terminates a bare operand. */
    private fun isOperandDelimiter(chars: String, end: Int): Boolean {
        val next = chars[end]
        if (next.isWhitespace()) return true
        if (next == '(' || next == ')' || next == '"') return true
        if (next == '<' || next == '>') return true
        if (end + 1 < chars.length) {
            val pair = chars.substring(end, end + 2)
            if (pair == "&&" || pair == "||" ||
                pair == "==" || pair == "!=" ||
                pair == "<=" || pair == ">="
            ) {
                return true
            }
        }
        return false
    }

    // MARK: - Parser entry point

    private fun parseToAST(expression: String): Node {
        val tokens = tokenize(expression)
        if (tokens.isEmpty()) {
            throw SimulationException(
                SimulationError.ScenarioValidationFailed(
                    "Condition expression '$expression' is empty",
                ),
            )
        }
        val parser = Parser(tokens, expression)
        val node = parser.parseOr()
        if (parser.position < tokens.size) {
            throw SimulationException(
                SimulationError.ScenarioValidationFailed(
                    "Condition expression '$expression' has unexpected trailing token " +
                        "near position ${parser.position} (likely an unmatched ')' or extra operator)",
                ),
            )
        }
        return node
    }

    /**
     * Recursive-descent parser. [position] is the cursor into [tokens]; each
     * `parseX` advances it past the tokens it consumed.
     */
    private class Parser(
        private val tokens: List<Token>,
        private val source: String,
    ) {
        var position: Int = 0

        private fun peek(): Token? = if (position < tokens.size) tokens[position] else null

        private fun advance() {
            position += 1
        }

        fun parseOr(): Node {
            var lhs = parseAnd()
            while (peek() == Token.LogicalOr) {
                advance()
                val rhs = parseAnd()
                lhs = Node.LogicalOr(lhs, rhs)
            }
            return lhs
        }

        private fun parseAnd(): Node {
            var lhs = parseFactor()
            while (peek() == Token.LogicalAnd) {
                advance()
                val rhs = parseFactor()
                lhs = Node.LogicalAnd(lhs, rhs)
            }
            return lhs
        }

        private fun parseFactor(): Node {
            if (peek() == Token.OpenParen) {
                advance()
                // Empty parens `()` — guard so the message is specific rather than
                // the generic "expected operand (got ')')" from parseComparison.
                if (peek() == Token.CloseParen) {
                    throw SimulationException(
                        SimulationError.ScenarioValidationFailed(
                            "Condition expression '$source' has empty parentheses '()'",
                        ),
                    )
                }
                val inner = parseOr()
                if (peek() != Token.CloseParen) {
                    throw SimulationException(
                        SimulationError.ScenarioValidationFailed(
                            "Condition expression '$source' is missing ')' to close a " +
                                "parenthesized sub-expression",
                        ),
                    )
                }
                advance()
                return inner
            }
            return parseComparison()
        }

        private fun parseComparison(): Node {
            val lhsToken = peek()
            if (lhsToken !is Token.Operand) {
                throw SimulationException(
                    SimulationError.ScenarioValidationFailed(
                        "Condition expression '$source' expected an operand ${describeUnexpected()}",
                    ),
                )
            }
            advance()
            val opToken = peek()
            if (opToken !is Token.ComparisonOp) {
                throw SimulationException(
                    SimulationError.ScenarioValidationFailed(
                        "Condition expression '$source' expected a comparison operator " +
                            "(==, !=, <, <=, >, >=) after '${lhsToken.text}' ${describeUnexpected()}",
                    ),
                )
            }
            advance()
            val rhsToken = peek()
            if (rhsToken !is Token.Operand) {
                throw SimulationException(
                    SimulationError.ScenarioValidationFailed(
                        "Condition expression '$source' expected an operand after " +
                            "'${lhsToken.text} ${opToken.symbol}' ${describeUnexpected()}",
                    ),
                )
            }
            advance()
            return Node.Comparison(lhsToken.text, opToken.symbol, rhsToken.text)
        }

        /** Short " (got 'X')" hint for the current token, or end-of-expression. */
        private fun describeUnexpected(): String =
            when (val token = peek()) {
                null -> "(reached end of expression)"
                Token.OpenParen -> "(got '(')"
                Token.CloseParen -> "(got ')')"
                Token.LogicalAnd -> "(got '&&')"
                Token.LogicalOr -> "(got '||')"
                is Token.ComparisonOp -> "(got '${token.symbol}')"
                is Token.Operand -> "(got '${token.text}')"
            }
    }

    private companion object {
        /**
         * A plain optionally-signed decimal with optional fraction / exponent.
         * Used for numeric detection in both [resolve] and [compare] so the
         * numeric-vs-string branch is identical across Swift and Kotlin — see the
         * class doc note on cross-language numeric parity.
         */
        private val numericLiteralRegex = Regex("^[+-]?[0-9]+(\\.[0-9]+)?([eE][+-]?[0-9]+)?$")

        private fun isNumericLiteral(token: String): Boolean = numericLiteralRegex.matches(token)

        private fun numericValue(token: String): Double? =
            if (isNumericLiteral(token)) token.toDouble() else null
    }
}
